engine/src/state_chain_observer/client.rs

```diff
@@ -279,52 +295,75 @@ impl<BaseRpcClient: base_rpc_api::BaseRpcApi + Send + Sync + 'static, SignedExtr
 			let mut block_stream =
 				process_stream_fn(Box::pin(block_stream.make_try_cached(latest_block))).await?;
 
-			let latest_block = *block_stream.cache();
-			if wait_for_required_version {
-				let incompatible_blocks =
-					futures::stream::once(futures::future::ready(Ok::<_, anyhow::Error>(
-						latest_block,
-					)))
-					.chain(block_stream.by_ref())
-					.try_take_while(|block_info| {
-						let base_rpc_client = base_rpc_client.clone();
-						let block_info = *block_info;
-						async move {
-							Ok(
-								if let Err(block_version) = base_rpc_client
-									.check_block_compatibility(block_info.hash)
-									.await?
-								{
-									info!(
-										"{} WAITING for a compatible release version.",
-										incompatible_block_message(block_version, block_info)
-									);
-									true
-								} else {
-									false
-								},
-							)
+			let (cfe_compatibility, latest_version_runtime_requires) =
+				base_rpc_client.check_block_compatibility(latest_block.hash).await?;
+
+			match cfe_compatibility {
+				CfeCompatibility::NoLongerCompatible => {
+					return Err(CreateBlockStreamError::NoLongerCompatible {
+						cfe_version: *CFE_VERSION,
+						cfe_version_required: latest_version_runtime_requires,
+						first_incompatible_block: latest_block.number,
+					}
+					.into());
+				},
+				CfeCompatibility::NotYetCompatible => {
+					if wait_for_required_version {
+						let incompatible_blocks = futures::stream::once(futures::future::ready(
+							Ok::<_, anyhow::Error>(latest_block),
+						))
+						.chain(block_stream.by_ref())
+						.try_take_while(|block_info| {
+							let base_rpc_client = base_rpc_client.clone();
+							let block_info = *block_info;
+							async move {
+								Ok({
+									let (cfe_compatibility, version_runtime_requires) =
+										base_rpc_client
+											.check_block_compatibility(block_info.hash)
+											.await?;
+
+									match cfe_compatibility {
+										CfeCompatibility::Compatible => false,
+										// We want the stream to continue as before
+										CfeCompatibility::NotYetCompatible => {
+											info!(
+												"{} WAITING for a compatible release version.",
+												incompatible_block_message(
```

Comment:
I forgot to inline this - I don't think it makes sense now to log it in the other cases, since we're just exiting, and the error will be logged at the top level